File: omrchecker-js/packages/core/src/processors/detection/bubbles_threshold/interpretation.ts
/**
 * Bubbles threshold interpretation.
 *
 * TypeScript port of src/processors/detection/bubbles_threshold/interpretation.py
 * Simplified bubble field interpretation using threshold strategies.
 */

import { Logger } from '../../../utils/logger';
import { FieldInterpretation, type InterpretationDrawing } from '../base/interpretation';
import type { Field } from '../../layout/field/base';
import {
  BubbleFieldDetectionResult,
  BubbleMeanValue,
} from '../models/detectionResults';
import { type ThresholdConfig, type ThresholdResult } from '../../threshold/GlobalThreshold';
import { LocalThreshold } from '../../threshold/LocalThreshold';
import type { BubblesScanBox } from '../../layout/field/bubbleField';
import { BubblesFieldInterpretationDrawing } from './interpretationDrawing';
import type { InterpretationThresholdConfig } from '../../../schemas/models/config';

const logger = new Logger('BubblesFieldInterpretation');

/**
 * Single bubble interpretation result.
 */
export class BubbleInterpretation {
  public bubbleMean: BubbleMeanValue;
  public threshold: number;
  public meanValue: number;
  public isAttempted: boolean;
  public bubbleValue: string;
  public itemReference: BubblesScanBox;

  constructor(bubbleMean: BubbleMeanValue, threshold: number) {
    this.bubbleMean = bubbleMean;
    this.threshold = threshold;
    this.meanValue = bubbleMean.meanValue;
    this.isAttempted = bubbleMean.meanValue < threshold;

    // Extract bubble value from unit_bubble if available
    // Note: unitBubble in BubbleMeanValue is BubbleLocation (interface), but in practice
    // it's a BubblesScanBox instance. We need to handle this properly.
    const unitBubble = bubbleMean.unitBubble as unknown as BubblesScanBox;
    this.bubbleValue = unitBubble?.bubbleValue || ''; // Ensure empty string if undefined/null

    // item_reference is used by the drawing code
    this.itemReference = unitBubble;
  }

  /**
   * Get bubble value if marked.
   *
   * @returns Bubble value or empty string
   */
  getValue(): string {
    return this.isAttempted ? this.bubbleValue : '';
  }
}

/**
 * Clean, simplified bubble field interpretation using strategies.
 *
 * Threshold calculation delegated to threshold strategy classes.
 */
export class BubblesFieldInterpretation extends FieldInterpretation {
  // Initialize with defaults. Parent constructor will call runInterpretation() which will
  // set proper values, but TypeScript properties declared without initializers are undefined
  // during parent constructor execution.
  public bubbleInterpretations: BubbleInterpretation[] = [];
  public isMultiMarked: boolean = false;
  public localThresholdForField: number = 0.0;
  public thresholdResult: ThresholdResult | null = null;

  constructor(
    tuningConfig: Record<string, unknown>,
    field: Field,
    fileLevelDetectionAggregates: unknown,
    fileLevelInterpretationAggregates: unknown
  ) {
    // Parent constructor calls runInterpretation(), but our property initializers
    // haven't run yet (they run AFTER super returns in TypeScript).
    // So runInterpretation sees undefined properties and initializes them.
    // Then when super() returns, our property initializers run and overwrite them.
    // Solution: Don't use property initializers, or re-run initialization after super().
    super(tuningConfig, field, fileLevelDetectionAggregates, fileLevelInterpretationAggregates);
    
    // Re-run interpretation after super() to ensure values are set after property initializers
    this.runInterpretation(field, fileLevelDetectionAggregates, fileLevelInterpretationAggregates);
  }

  /**
   * Get drawing instance for visualization.
   */
  getDrawingInstance(): InterpretationDrawing {
    return new BubblesFieldInterpretationDrawing(this);
  }

  /**
   * Get final interpretation string.
   *
   * Returns concatenated marked bubble values or empty value.
   * Special case: If ALL bubbles are marked, treat as unmarked (likely scanning issue).
   *
   * @returns Final interpretation string
   */
  getFieldInterpretationString(): string {
    const markedBubbles = this.bubbleInterpretations
      .filter((interp) => interp.isAttempted)
      .map((interp) => interp.bubbleValue);

    // If no bubbles marked, return empty value
    if (markedBubbles.length === 0) {
      return this.emptyValue;
    }

    // If ALL bubbles are marked, treat as unmarked (likely scanning/detection issue)
    const totalBubbles = this.bubbleInterpretations.length;
    if (markedBubbles.length === totalBubbles) {
      return this.emptyValue;
    }

    return markedBubbles.join('');
  }

  /**
   * Run interpretation using detection results and threshold strategies.
   *
   * @param field - Field to interpret
   * @param fileLevelDetectionAggregates - Detection results
   * @param fileLevelInterpretationAggregates - Interpretation aggregates
   */
  runInterpretation(
    field: Field,
    fileLevelDetectionAggregates: unknown,
    fileLevelInterpretationAggregates: unknown
  ): void {
    // Initialize properties first to ensure they're never undefined
    this.bubbleInterpretations = [];
    this.isMultiMarked = false;
    this.localThresholdForField = 0.0;
    this.thresholdResult = null;

    // Step 1: Extract detection result
    const detectionResult = this.extractDetectionResult(
      field,
      fileLevelDetectionAggregates
    );

    if (!detectionResult) {
      const aggregates = fileLevelDetectionAggregates as Record<string, unknown>;
      const bubbleFields = aggregates.bubble_fields as Record<string, unknown> | undefined;
      const availableFields = bubbleFields ? Object.keys(bubbleFields) : [];
      
      throw new Error(
        `No detection result for field '${field.fieldLabel}'. Available: [${availableFields.join(', ')}]`
      );
    }
    if (!detectionResult.bubbleMeans || detectionResult.bubbleMeans.length === 0) {
      throw new Error(
        `No bubble means in detection result for field '${field.fieldLabel}'`
      );
    }

    // Step 2: Calculate thresholds using strategies
    const thresholdConfig = this.createThresholdConfig(
      fileLevelInterpretationAggregates
    );

    const result = this.calculateThreshold(
      detectionResult,
      fileLevelInterpretationAggregates,
      thresholdConfig
    );

    if (!result) {
      logger.error('calculateThreshold returned null/undefined');
      return;
    }

    this.thresholdResult = result;
    this.localThresholdForField = result.thresholdValue;

    // Step 3: Interpret bubbles
    this.interpretBubbles(detectionResult);

    // Step 4: Check multi-marking
    this.checkMultiMarking();

    // Step 5: Calculate confidence metrics (if enabled)
    const outputs = this.tuningConfig.outputs as { show_confidence_metrics?: boolean } | undefined;
    if (outputs?.show_confidence_metrics) {
      this.calculateConfidenceMetrics(
        detectionResult,
        fileLevelInterpretationAggregates
      );
    }
  }

  /**
   * Extract detection result from aggregates.
   *
   * Can work with both new typed models and legacy dict format.
   */
  private extractDetectionResult(
    field: Field,
    fileLevelDetectionAggregates: unknown
  ): BubbleFieldDetectionResult | null {
    const fieldLabel = field.fieldLabel;
    const aggregates = fileLevelDetectionAggregates as Record<string, unknown>;

    // Try to get from new typed format first
    if ('bubble_fields' in aggregates) {
      const bubbleFields = aggregates.bubble_fields as Record<string, BubbleFieldDetectionResult>;
      if (bubbleFields[fieldLabel]) {
        return bubbleFields[fieldLabel];
      }
      logger.warn(`Field ${fieldLabel} not found in bubble_fields`);
      return null;
    }

    // Fallback to legacy dict format
    const fieldLabelWiseAggregates = aggregates.field_label_wise_aggregates as Record<
      string,
      { field_bubble_means: BubbleMeanValue[] }
    >;

    if (!fieldLabelWiseAggregates || !fieldLabelWiseAggregates[fieldLabel]) {
      logger.warn(
        `bubble_fields not found in file_level_detection_aggregates for field ${fieldLabel}`
      );
      return null;
    }

    const fieldLevelDetectionAggregates = fieldLabelWiseAggregates[fieldLabel];

    // Create typed result from legacy format
    return new BubbleFieldDetectionResult(
      field.id,
      fieldLabel,
      fieldLevelDetectionAggregates.field_bubble_means
    );
  }

  /**
   * Create threshold configuration from tuning config and file-level aggregates.
   */
  private createThresholdConfig(
    _fileLevelInterpretationAggregates: unknown
  ): ThresholdConfig {
    const thresholding = this.tuningConfig.thresholding as InterpretationThresholdConfig;

    return {
      defaultThreshold: thresholding?.global_page_threshold || 180,
      minJump: thresholding?.min_jump || 10,
      minGapTwoBubbles: thresholding?.min_gap_two_bubbles || 20,
      minJumpSurplusForGlobalFallback: thresholding?.min_jump_surplus_for_global_fallback || 10,
      // Note: TypeScript ThresholdConfig is simpler than Python's
      // Additional config options can be added if needed
    };
  }

  /**
   * Calculate threshold using strategies.
   *
   * Uses LocalThreshold with global fallback.
   */
  private calculateThreshold(
    detectionResult: BubbleFieldDetectionResult,
    fileLevelInterpretationAggregates: unknown,
    config: ThresholdConfig
  ): ThresholdResult {
    // Get global fallback threshold
    const fileLevelAggs = fileLevelInterpretationAggregates as Record<string, unknown>;
    const globalFallback =
      (fileLevelAggs.file_level_fallback_threshold as number) || config.defaultThreshold;

    // Use local strategy with global fallback
    const localThreshold = new LocalThreshold(globalFallback);
    const meanValues = detectionResult.meanValues;

    // Calculate threshold
    return localThreshold.calculateThreshold(meanValues, config);
  }

  /**
   * Interpret each bubble using calculated threshold.
   *
   * Creates interpretation for each bubble.
   */
  private interpretBubbles(detectionResult: BubbleFieldDetectionResult): void {
    if (!detectionResult || !detectionResult.bubbleMeans) {
      logger.warn('No bubble means in detection result');
      this.bubbleInterpretations = [];
      return;
    }
    this.bubbleInterpretations = detectionResult.bubbleMeans.map(
      (bubbleMean) =>
        new BubbleInterpretation(bubbleMean, this.localThresholdForField)
    );
  }

  /**
   * Check if multiple bubbles are marked.
   */
  private checkMultiMarking(): void {
    const markedCount = this.bubbleInterpretations.filter(
      (interp) => interp.isAttempted
    ).length;
    this.isMultiMarked = markedCount > 1;

    if (this.isMultiMarked) {
      logger.warn(
        `Multi-marking detected in field: ${this.field.fieldLabel}, marked bubbles: ${markedCount}`
      );
    }
  }

  /**
   * Calculate confidence metrics for the field.
   *
   * Simplified version - can be expanded based on threshold result.
   */
  private calculateConfidenceMetrics(
    detectionResult: BubbleFieldDetectionResult,
    fileLevelAggregates: unknown
  ): void {
    const aggregates = fileLevelAggregates as Record<string, unknown>;
    const globalThreshold =
      (aggregates.file_level_fallback_threshold as number) ||
      this.thresholdResult?.thresholdValue ||
      180;

    // Check for disparity between global and local thresholds
    const disparityBubbles: BubbleMeanValue[] = [];
    for (const bubbleMean of detectionResult.bubbleMeans) {
      const localMarked = bubbleMean.meanValue < this.localThresholdForField;
      const globalMarked = bubbleMean.meanValue < globalThreshold;

      if (localMarked !== globalMarked) {
        disparityBubbles.push(bubbleMean);
      }
    }

    // Calculate overall confidence score for ML training
    const confidenceScore = this.calculateOverallConfidenceScore(
      detectionResult,
      disparityBubbles
    );

    // Build confidence metrics
    this.insertFieldLevelConfidenceMetrics({
      local_threshold: this.localThresholdForField,
      global_threshold: globalThreshold,
      threshold_confidence: this.thresholdResult?.confidence || 0,
      threshold_method: this.thresholdResult?.methodUsed || 'unknown',
      max_jump: this.thresholdResult?.maxJump || 0,
      bubbles_in_doubt: {
        by_disparity: disparityBubbles,
      },
      is_local_jump_confident: (this.thresholdResult?.confidence || 0) > 0.7,
      field_label: this.field.fieldLabel,
      scan_quality: detectionResult.scanQuality,
      std_deviation: detectionResult.stdDeviation,
      overall_confidence_score: confidenceScore,
    });

    if (disparityBubbles.length > 0) {
      logger.warn(
        `Threshold disparity in field: ${this.field.fieldLabel}, bubbles in doubt: ${disparityBubbles.length}`
      );
    }
  }

  /**
   * Calculate overall confidence score for this field's detection.
   *
   * Score ranges from 0.0 to 1.0 based on:
   * - Threshold margin (how far marked bubbles are from threshold)
   * - Multi-mark probability
   * - Bubble intensity consistency within field
   * - Disparity with global threshold
   *
   * @returns Confidence score between 0.0 (low confidence) and 1.0 (high confidence)
   */
  private calculateOverallConfidenceScore(
    detectionResult: BubbleFieldDetectionResult,
    disparityBubbles: BubbleMeanValue[]
  ): number {
    if (detectionResult.bubbleMeans.length === 0) {
      return 0.0;
    }

    if (!this.thresholdResult) {
      return 0.0;
    }

    // Factor 1: Threshold confidence from strategy (0.0-1.0)
    const thresholdConfidence = this.thresholdResult.confidence;

    // Factor 2: Margin from threshold (how clearly marked/unmarked)
    const markedBubbles = detectionResult.bubbleMeans.filter(
      (b) => b.meanValue < this.localThresholdForField
    );

    let marginConfidence: number;
    if (markedBubbles.length > 0) {
      const avgMargin =
        markedBubbles.reduce(
          (sum, b) => sum + (this.localThresholdForField - b.meanValue),
          0
        ) / markedBubbles.length;
      // Normalize margin confidence (larger margin = higher confidence)
      // Assume 50 intensity units is very confident
      marginConfidence = Math.min(1.0, avgMargin / 50.0);
    } else {
      // No bubbles marked - check unmarked confidence
      const avgDistance =
        detectionResult.bubbleMeans.reduce(
          (sum, b) => sum + (b.meanValue - this.localThresholdForField),
          0
        ) / detectionResult.bubbleMeans.length;
      marginConfidence = Math.min(1.0, avgDistance / 50.0);
    }

    // Factor 3: Multi-mark penalty
    const markedCount = markedBubbles.length;
    let multiMarkPenalty: number;
    if (markedCount > 1) {
      multiMarkPenalty = 0.3; // Reduce by 30%
    } else if (markedCount === 0) {
      multiMarkPenalty = 0.1; // Slight penalty
    } else {
      multiMarkPenalty = 0.0; // Single mark - ideal
    }

    // Factor 4: Disparity penalty
    const disparityRatio =
      detectionResult.bubbleMeans.length > 0
        ? disparityBubbles.length / detectionResult.bubbleMeans.length
        : 0;
    const disparityPenalty = disparityRatio * 0.4; // Up to 40% penalty

    // Factor 5: Scan quality
    const scanQualityMap: Record<string, number> = {
      EXCELLENT: 1.0,
      GOOD: 0.9,
      MODERATE: 0.7,
      POOR: 0.5,
    };
    const scanQualityFactor =
      scanQualityMap[detectionResult.scanQuality] || 0.5;

    // Combine factors (weighted average)
    const confidenceScore =
      (thresholdConfidence * 0.35 +
        marginConfidence * 0.25 +
        scanQualityFactor * 0.2) *
      (1.0 - multiMarkPenalty - disparityPenalty);

    // Clamp to [0, 1]
    return Math.max(0.0, Math.min(1.0, confidenceScore));
  }
}

File: src/processors/detection/bubbles_threshold/interpretation.py
from src.processors.detection.base.interpretation import FieldInterpretation
from src.processors.detection.bubbles_threshold.interpretation_drawing import (
    BubblesFieldInterpretationDrawing,
)
from src.processors.detection.models.detection_results import (
    BubbleFieldDetectionResult,
)
from src.processors.layout.field.base import Field
from src.processors.threshold.local_threshold import LocalThresholdStrategy
from src.processors.threshold.threshold_result import ThresholdConfig, ThresholdResult
from src.utils.logger import logger


class BubbleInterpretation:
    """Single bubble interpretation result."""

    def __init__(self, bubble_mean, threshold: float) -> None:
        self.bubble_mean = bubble_mean
        self.threshold = threshold
        self.mean_value = bubble_mean.mean_value
        self.is_attempted = bubble_mean.mean_value < threshold
        self.bubble_value = (
            bubble_mean.unit_bubble.bubble_value
            if hasattr(bubble_mean.unit_bubble, "bubble_value")
            else ""
        )
        # item_reference is used by the drawing code
        self.item_reference = bubble_mean.unit_bubble

    def get_value(self) -> str:
        """Get bubble value if marked."""
        return self.bubble_value if self.is_attempted else ""


class BubblesFieldInterpretation(FieldInterpretation):
    def __init__(self, *args, **kwargs) -> None:
        self.bubble_interpretations: list[BubbleInterpretation] = []
        self.is_multi_marked = False
        self.local_threshold_for_field = 0.0
        self.threshold_result: ThresholdResult | None = None
        super().__init__(*args, **kwargs)

    def get_drawing_instance(self):
        """Get drawing instance for visualization."""
        return BubblesFieldInterpretationDrawing(self)

    def get_field_interpretation_string(self) -> str:
        """Get final interpretation string.

        Returns concatenated marked bubble values or empty value.
        Special case: If ALL bubbles are marked, treat as unmarked (likely scanning issue).
        """
        marked_bubbles = [
            interp.bubble_value
            for interp in self.bubble_interpretations
            if interp.is_attempted
        ]

        # If no bubbles marked, return empty value
        if len(marked_bubbles) == 0:
            return self.empty_value

        # If ALL bubbles are marked, treat as unmarked (likely scanning/detection issue)
        total_bubbles = len(self.bubble_interpretations)
        if len(marked_bubbles) == total_bubbles:
            return self.empty_value

        return "".join(marked_bubbles)

    def run_interpretation(
        self,
        field: Field,
        file_level_detection_aggregates,
        file_level_interpretation_aggregates,
    ) -> None:
        """Run interpretation using detection results and threshold strategies.

        Args:
            field: Field to interpret
            file_level_detection_aggregates: Detection results (typed models via bubble_fields)
            file_level_interpretation_aggregates: Interpretation aggregates
        """
        # Step 1: Extract detection result
        detection_result = self._extract_detection_result(
            field, file_level_detection_aggregates
        )

        # Step 2: Calculate thresholds using strategies
        threshold_config = self._create_threshold_config(
            file_level_interpretation_aggregates
        )
        self.threshold_result = self._calculate_threshold(
            detection_result, file_level_interpretation_aggregates, threshold_config
        )

        self.local_threshold_for_field = self.threshold_result.threshold_value

        # Step 3: Interpret bubbles
        self._interpret_bubbles(detection_result)

        # Step 4: Check multi-marking
        self._check_multi_marking()

        # Step 5: Calculate confidence metrics (if enabled)
        if self.tuning_config.outputs.show_confidence_metrics:
            self._calculate_confidence_metrics(
                detection_result, file_level_interpretation_aggregates
            )

    def _extract_detection_result(
        self, field: Field, file_level_detection_aggregates
    ) -> BubbleFieldDetectionResult:
        """Extract detection result from aggregates.

        Uses new typed models pipeline via bubble_fields.
        """
        field_label = field.field_label

        # Get from new typed format
        if "bubble_fields" not in file_level_detection_aggregates:
            msg = f"bubble_fields not found in file_level_detection_aggregates for field {field_label}"
            raise KeyError(msg)

        bubble_fields = file_level_detection_aggregates["bubble_fields"]
        if field_label not in bubble_fields:
            msg = f"Field {field_label} not found in bubble_fields"
            raise KeyError(msg)

        return bubble_fields[field_label]

    def _create_threshold_config(
        self, file_level_interpretation_aggregates
    ) -> ThresholdConfig:
        """Create threshold configuration from tuning config and file-level aggregates."""
        config = self.tuning_config
        return ThresholdConfig(
            min_jump=config.thresholding.MIN_JUMP,
            jump_delta=config.thresholding.JUMP_DELTA,
            min_gap_two_bubbles=config.thresholding.MIN_GAP_TWO_BUBBLES,
            min_jump_surplus_for_global_fallback=config.thresholding.MIN_JUMP_SURPLUS_FOR_GLOBAL_FALLBACK,
            confident_jump_surplus_for_disparity=config.thresholding.CONFIDENT_JUMP_SURPLUS_FOR_DISPARITY,
            global_threshold_margin=config.thresholding.GLOBAL_THRESHOLD_MARGIN,
            outlier_deviation_threshold=file_level_interpretation_aggregates.get(
                "outlier_deviation_threshold_for_file", 5.0
            ),
            default_threshold=config.thresholding.GLOBAL_PAGE_THRESHOLD,
        )

    def _calculate_threshold(
        self,
        detection_result: BubbleFieldDetectionResult,
        file_level_interpretation_aggregates,
        config: ThresholdConfig,
    ) -> ThresholdResult:
        """Calculate threshold using strategies.

        Uses LocalThresholdStrategy with global fallback.
        This replaces 170+ lines of threshold calculation code!
        """
        # Get global fallback threshold
        global_fallback = file_level_interpretation_aggregates.get(
            "file_level_fallback_threshold", config.default_threshold
        )

        # Use local strategy with global fallback
        strategy = LocalThresholdStrategy(global_fallback=global_fallback)

        # Calculate threshold
        return strategy.calculate_threshold(
            detection_result.mean_values,  # Auto-extracted from model!
            config,
        )

    def _interpret_bubbles(self, detection_result: BubbleFieldDetectionResult) -> None:
        """Interpret each bubble using calculated threshold.

        Creates interpretation for each bubble.
        """
        self.bubble_interpretations = [
            BubbleInterpretation(bubble_mean, self.local_threshold_for_field)
            for bubble_mean in detection_result.bubble_means
        ]

    def _check_multi_marking(self) -> None:
        """Check if multiple bubbles are marked."""
        marked_count = sum(
            1 for interp in self.bubble_interpretations if interp.is_attempted
        )
        self.is_multi_marked = marked_count > 1

        if self.is_multi_marked:
            logger.warning(
                f"Multi-marking detected in field: {self.field.field_label}, "
                f"marked bubbles: {marked_count}"
            )

    def _calculate_confidence_metrics(
        self, detection_result: BubbleFieldDetectionResult, file_level_aggregates
    ) -> None:
        """Calculate confidence metrics for the field.

        Simplified version - can be expanded based on threshold result.
        """
        global_threshold = file_level_aggregates.get(
            "file_level_fallback_threshold", self.threshold_result.threshold_value
        )

        # Check for disparity between global and local thresholds
        disparity_bubbles = []
        for bubble_mean in detection_result.bubble_means:
            local_marked = bubble_mean.mean_value < self.local_threshold_for_field
            global_marked = bubble_mean.mean_value < global_threshold

            if local_marked != global_marked:
                disparity_bubbles.append(bubble_mean)

        # Calculate overall confidence score for ML training
        confidence_score = self._calculate_overall_confidence_score(
            detection_result, disparity_bubbles
        )

        # Build confidence metrics
        self.insert_field_level_confidence_metrics(
            {
                "local_threshold": self.local_threshold_for_field,
                "global_threshold": global_threshold,
                "threshold_confidence": self.threshold_result.confidence,
                "threshold_method": self.threshold_result.method_used,
                "max_jump": self.threshold_result.max_jump,
                "bubbles_in_doubt": {
                    "by_disparity": disparity_bubbles,
                },
                "is_local_jump_confident": self.threshold_result.confidence > 0.7,
                "field_label": self.field.field_label,
                "scan_quality": detection_result.scan_quality.value,
                "std_deviation": detection_result.std_deviation,
                "overall_confidence_score": confidence_score,  # NEW: For ML training
            }
        )

        if len(disparity_bubbles) > 0:
            logger.warning(
                f"Threshold disparity in field: {self.field.field_label}, "
                f"bubbles in doubt: {len(disparity_bubbles)}"
            )

    def _calculate_overall_confidence_score(
        self, detection_result: BubbleFieldDetectionResult, disparity_bubbles
    ) -> float:
        """Calculate overall confidence score for this field's detection.

        Score ranges from 0.0 to 1.0 based on:
        - Threshold margin (how far marked bubbles are from threshold)
        - Multi-mark probability
        - Bubble intensity consistency within field
        - Disparity with global threshold

        Returns:
            float: Confidence score between 0.0 (low confidence) and 1.0 (high confidence)
        """
        if not detection_result.bubble_means:
            return 0.0

        # Factor 1: Threshold confidence from strategy (0.0-1.0)
        threshold_confidence = self.threshold_result.confidence

        # Factor 2: Margin from threshold (how clearly marked/unmarked)
        # Calculate average distance from threshold for marked bubbles
        marked_bubbles = [
            b
            for b in detection_result.bubble_means
            if b.mean_value < self.local_threshold_for_field
        ]
        if marked_bubbles:
            avg_margin = sum(
                self.local_threshold_for_field - b.mean_value for b in marked_bubbles
            ) / len(marked_bubbles)
            # Normalize margin confidence (larger margin = higher confidence)
            # Assume 50 intensity units is very confident
            margin_confidence = min(1.0, avg_margin / 50.0)
        else:
            # No bubbles marked - check unmarked confidence
            avg_distance = sum(
                b.mean_value - self.local_threshold_for_field
                for b in detection_result.bubble_means
            ) / len(detection_result.bubble_means)
            margin_confidence = min(1.0, avg_distance / 50.0)

        # Factor 3: Multi-mark penalty
        marked_count = len(marked_bubbles)
        if marked_count > 1:
            # Multi-marking reduces confidence
            multi_mark_penalty = 0.3  # Reduce by 30%
        elif marked_count == 0:
            # No marks - still confident if thresholds agree
            multi_mark_penalty = 0.1  # Slight penalty
        else:
            # Single mark - ideal
            multi_mark_penalty = 0.0

        # Factor 4: Disparity penalty
        disparity_ratio = (
            len(disparity_bubbles) / len(detection_result.bubble_means)
            if detection_result.bubble_means
            else 0
        )
        disparity_penalty = disparity_ratio * 0.4  # Up to 40% penalty

        # Factor 5: Scan quality
        scan_quality_map = {
            "EXCELLENT": 1.0,
            "GOOD": 0.9,
            "MODERATE": 0.7,
            "POOR": 0.5,
        }
        scan_quality_factor = scan_quality_map.get(
            detection_result.scan_quality.value, 0.5
        )

        # Combine factors (weighted average)
        confidence_score = (
            threshold_confidence * 0.35
            + margin_confidence * 0.25
            + scan_quality_factor * 0.20
        ) * (1.0 - multi_mark_penalty - disparity_penalty)

        # Clamp to [0, 1]
        return max(0.0, min(1.0, confidence_score))
